mining: fix -blockreservedweight shadows IPC option

Sjors opened the pull request:
Also enforce `MINIMUM_BLOCK_RESERVED_WEIGHT` for IPC clients.

The `-blockreservedweight` startup option should only affect RPC code, because IPC clients (currently) do not have a way to signal their intent to use the node default (the `BlockCreateOptions` struct defaults merely document a recommendation for client software).

Before this PR however, if the user set `-blockreservedweight` then `ApplyArgsManOptions` would cause the `block_reserved_weight` option passed by IPC clients to be ignored. _Users who don't set this value were not affected._

Fix this by making BlockCreateOptions::block_reserved_weight an std::optional.

Internal interface users, such as the RPC call sites, don't set a value so -blockreservedweight is used. Whereas IPC clients do set a value which is no longer ignored.

Test coverage is added, with a preliminary commit that refactors the `create_block_template` and `wait_next_template` helpers.

`mining_basic.py` already ensured `-blockreservedweight` is enforced by mining RPC methods. The second commit adds coverage for Mining interface IPC clients. It also verifies that `-blockreservedweight` has no effect on them.

The third commit enforces `MINIMUM_BLOCK_RESERVED_WEIGHT` for IPC clients. Previously lower values were quietly clamped.

---

Merge order preference: #34452 should ideally go first.

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -376,7 +379,11 @@ static RPCHelpMan generateblock()
+            std::unique_ptr<BlockTemplate> block_template{miner.createNewBlock({
+                .use_mempool = false,
+                .coinbase_output_script = coinbase_output_script,
+                .block_reserved_weight = size_t(node.args->GetIntArg("-blockreservedweight", DEFAULT_BLOCK_RESERVED_WEIGHT))
```
I have a larger refactor coming soon that simplifies this to `node.mining_args.default_block_reserved_weight`. But this is easier to backport.

Sjors commented on src/rpc/mining.cpp:
```diff
@@ -871,7 +878,9 @@ static RPCHelpMan getblocktemplate()
         // Create new block
-        block_template = miner.createNewBlock();
+        block_template = miner.createNewBlock({
+            .block_reserved_weight = size_t(node.args->GetIntArg("-blockreservedweight", DEFAULT_BLOCK_RESERVED_WEIGHT))
```
I have a larger refactor coming soon that simplifies this to `node.mining_args.default_block_reserved_weight`. But this is easier to backport.

Sjors replied:
The latest push leaves RPC code untouched.

Sjors commented:
> if you think it would be useful for IPC clients to "have a way to signal their intent to use the node default"

sv2-tp always sets the value (based on `coinbase_output_max_additional_size` in  `CoinbaseOutputConstraints`), so it wouldn't benefit. But it does seem better to allow clients to omit this value.

For that to work however, we'd also need to change mining.capnp, e.g.:

```capnp
blockReservedWeight @1 :UInt64 $Proxy.name("block_reserved_weight");
hasBlockReservedWeight @3 :Bool = true $Proxy.skip;
```

Adding the `has` field and having it default to `true` means it's not a breaking change. But this doesn't actually work :-)

Also, if we really wanted to make this optional for IPC clients, I would prefer to add direct support for optionals to capnp files, as a breaking change.


However, I did take your patch because it allows this PR to simplified:
- internal callers will see `coinbase_output_max_additional_size` as optional and can omit it. That lets me avoid the four `GetIntArg` calls (and leave `rpc/mining.cpp` alone).
- IPC callers don't see it as optional and always provide a value (which we can change later)

I added documentation to make this behavior clear.

ryanofsky reviewed (approved):
Code review ACK 1f2f4ab9d3fbdd3071810fa78a17563df3f56f91. This is a pretty clean way of letting the IPC reserved weight value take precedence over the `-blockreservedweight`, and a good first step towards making the IPC value optional in the future.

re: https://github.com/bitcoin/bitcoin/pull/33965#issuecomment-3601375713

> For that to work however, we'd also need to change mining.capnp, e.g.:

Yes, sorry, I should have mentioned that. I think it makes sense to add support in the c++ implementation first, and the capn'proto interface later. There are many ways the capnproto interface could be modified to support this and be backwards compatible if desired. I think the nicest way to support making the value optional might be to use a union like:

```
blockReservedWeight : union { none @1 :Void; value @2 :UInt64; };
```

(Interestingly according to https://capnproto.org/language.html#unions, you can even convert a non-union field into a union field without breaking backwards compatibility, as long as the old field is the lowest-numbered field in the union. However, this numbering would also affect the union default value so probably not worth it.)

Another way to support making the value optional would be to add a boolean `has` field like you suggested. Another way would be to "box" the value, putting it in a struct with a single field. Probably all approaches will require modifying libmultiprocess somehow, so would be better to implement in a separate PR.

Sjors commented on src/node/types.h:
```diff
@@ -41,8 +42,10 @@ struct BlockCreateOptions {
      * The default reserved weight for the fixed-size block header,
      * transaction count and coinbase transaction.
+     *
+     * Not optional for IPC clients.
```
I took the suggested message.

ryanofsky reviewed (approved):
Code review ACK 4d686142e079765cd31851481deb70659a9e9376. Just updated a comment since last review. Still looks good! Only minor comments below, not important

ryanofsky reviewed (approved):
Code review ACK 54d3054db129d4dd5aaf9faa3f6e6922159059b0. Just some more review suggestions applied since last review. This is a pretty clean fix preventing the `-blockreservedweight` configuration value from overriding the IPC requested value.

The followup #33966 will make the code merging IPC options and configuration options more clear. And a different followup could give IPC clients ability to use the `-blockreservedweight` value instead of needing to provide their own value.

Sjors commented:
Rebased after #34003. I added an initial test refactor commit to have the `{create_block,wait_next}_template` helpers introduced in that PR to return `None` if there's a timeout or failure.

I also added a commit to enforce (rather than clamp) the minimum reserved block value, as noticed by @plebhash in https://github.com/stratum-mining/sv2-tp/pull/51#issuecomment-3670613634. I could make that it's own PR but it's rather trivial.

maflcko commented on src/node/interfaces.cpp:
```diff
@@ -963,6 +963,11 @@ class MinerImpl : public Mining
 
     std::unique_ptr<BlockTemplate> createNewBlock(const BlockCreateOptions& options) override
     {
+        // Enfore minimum block_reserved_weight to avoid silently clamping it
```
from the LLM:


    Enfore -> Enforce [misspelling in comment "Enfore minimum block_reserved_weight..." makes the word invalid]
    wait -> weight [comment "Enforce minimum reserved wait for IPC clients too" should say "reserved weight" to match intended meaning]

Sjors replied:
How many LLM's does it take to fix a typo? :-)
<img width="525" height="649" alt="Scherm­afbeelding 2025-12-19 om 16 59 50" src="https://github.com/user-attachments/assets/4e8bb667-3217-4dc1-a74e-b83410114ba4" />

Sjors committed:
test: have mining template helpers return None

Refactor the mining_create_block_template and mining_wait_next_template
helpers in ipc_util.py to return None if they time out or fail. It makes
the test easier to read and provides a more clear error message in case
of a regression.

There were a few spots that didn't use mining_wait_next_template yet,
which now do.

Sjors committed:
mining: fix -blockreservedweight shadows IPC option

The -blockreservedweight startup option should only affect RPC code,
because IPC clients (currently) do not have a way to signal their intent
to use the node default (the BlockCreateOptions struct defaults
merely document a recommendation for client software).

Before this commit however, if the user set -blockreservedweight
then ApplyArgsManOptions would cause the block_reserved_weight
option passed by IPC clients to be ignored. Users who don't set
this value were not affected.

Fix this by making BlockCreateOptions::block_reserved_weight an
std::optional.

Internal interface users, such as the RPC call sites, don't set a
value so -blockreservedweight is used. Whereas IPC clients do set
a value which is no longer ignored.

Test coverage is added.

mining_basic.py already ensured -blockreservedweight is enforced by
mining RPC methods. This commit adds coverage for Mining interface IPC
clients. It also verifies that -blockreservedweight has no effect on
them.

Co-Authored-By: Russell Yanofsky <russ@yanofsky.org>

plebhash commented:
> @plebhash maybe you can you give this a quick test against the SRI integration? Without this PR, if you set `-blockreservedweigh` it will ignore the corresponding IPC argument.

what methodology would you suggest for this?

I assume we want to mine a block that's filled up to its capacity, and then check the consumed weight to see how far it is from the consensus limit, right?

I'm trying to leverage `integration_tests_sv2` crate, which allows us to write some Rust code to interact with a regtest node

but I'm having a really hard time to reliably populate the mempool with enough transactions to fill an entire block

Sjors commented:
@plebhash in the functional test for this PR I simply set `-blockreservedweight=4000000`. That should result in empty blocks. Then just use a normal value in the IPC calls and the blocks will still be empty (before this PR).

Sjors commented:
@plebhash thanks for testing! So it looks like this PR works.

> this makes me conclude that Bitcoin Core will not accept a `CreateNewBlock` with `set_block_reserved_weight < 2000`, which feels aligned with the rationale behind [stratum-mining/sv2-tp#79](https://github.com/stratum-mining/sv2-tp/pull/79)

Good to know that it's not quietly increasing the value, but rather just fails. Even though the error could be better.

ryanofsky commented on src/node/types.h:
```diff
@@ -41,7 +41,7 @@ struct BlockCreateOptions {
     /**
      * The default reserved weight for the fixed-size block header,
-     * transaction count and coinbase transaction.
+     * transaction count and coinbase transaction. Minimum: 2000 weight units.
```
In commit "mining: fix -blockreservedweight shadows IPC option" (336826ad0e70978783dcc0777ad1f01d3dbc1698)

Maybe update the comment to say "2000 weight units (MINIMUM_BLOCK_RESERVED_WEIGHT)". This would be helpful when grepping the constant and seeing how it is used & what may need to be updated if it changes.

ryanofsky reviewed (approved):
Code review ACK 31325c1782e542601b3de8bd8bafe0af372830e9

Many changes since last review.

Main code change in second commit is a straightforward fix using `std::optional` to prevent `ApplyArgsManOptions` from overriding the `block_reserved_weight` value provided by IPC and ignoring it.

The test improvements in the first commit are also nice, and the change in the third commit to enforce the MINIMUM_BLOCK_RESERVED_WEIGHT instead of silently clamping it seem good.

I left some suggestions below but none are important.

Sjors commented:
Addressed comments and also rebased (needed for #33966).

ryanofsky reviewed (approved):
Code review ACK d3db5285989f7b3fef7436a37ce928065be2804e just making reserved weight check into an exception and improving comments since last review. Thanks for the updates!

ryanofsky reviewed (approved):
Code review ACK d443301666ba4ad4fbf5388c85862a6cae290d53, just catching more specific python exception in test since last review

ryanofsky reviewed (approved):
Code review ACK 8bc1cb77fada68796c9599a04a66025a4b53ddab, just rebased with include/import changes since last review

sedited commented on src/node/miner.cpp:
```diff
@@ -102,13 +103,15 @@ void ApplyArgsManOptions(const ArgsManager& args, BlockAssembler::Options& optio
 {
     // Reserve space for fixed-size block header, txs count, and coinbase tx.
-    nBlockWeight = m_options.block_reserved_weight;
+    nBlockWeight = *m_options.block_reserved_weight;
```
Nit: Should we sprinkle in an `Assert(` here?

sedited reviewed (approved):
ACK 8bc1cb77fada68796c9599a04a66025a4b53ddab

ismaelsadeeq commented on src/node/miner.cpp:
```diff
@@ -78,11 +78,12 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman)
 static BlockAssembler::Options ClampOptions(BlockAssembler::Options options)
 {
-    options.block_reserved_weight = std::clamp<size_t>(options.block_reserved_weight, MINIMUM_BLOCK_RESERVED_WEIGHT, MAX_BLOCK_WEIGHT);
+    // Apply DEFAULT_BLOCK_RESERVED_WEIGHT when the caller left it unset.
```
In "mining: fix -blockreservedweight shadows IPC option" f69320d3279e2abe1e5d47f496ad54deb4c20fcf

Why make the distinction between RPC and IPC clients?

The narrative of this commit should rather be future proof and changed to: it can be overridded by runtime via the IPC or RPC.

ismaelsadeeq reviewed (commented):
Code review ACK 8bc1cb77fada68796c9599a04a66025a4b53ddab

I verified that we now make `block_reserved_weight` optional, such that whenever a value is provided at runtime, that value is used; it is not silently changed to the startup value in `ApplyArgsManOptions`.

I also verify that the current change does not allow for a default or overridden `blockreservedweight` set via startup to be lower (when the rpc is used), by generating a block template and ensuring its weight does not go beyond `4000000` - `-blockreservedweight`.

IMO, the way the current functionality is implemented now using `std::optional` is a bit awkward, and not going to be easy to extend. for e.g if IPC has other options in the future that get set at runtime, and we don't want to override like `blockmaxweight`? Would those also need to be wrapped in `std::optional`? 

A better solution is to split `ApplyArgsManOptions` into two functions, instead `ApplyMandatoryOptions`  for mandatory overrides that must be called, and  `ApplyStartupOptions`  for startup-specific options that only mining RPCs call, since they don't have runtime overrides.  IPC would only call `ApplyMandatoryOptions`. 

Alternatively, we could also use a boolean parameter in the options struct so that `ApplyArgsManOptions` can skip certain overrides altogether, which is also easy to extend, but booleans can be brittle.

ryanofsky commented on src/node/interfaces.cpp:
```diff
@@ -969,6 +970,16 @@ class MinerImpl : public Mining
+        // end up mining with different options than requested. This matches the
+        // behavior of the `-blockreservedweight` startup option, which rejects
+        // values below MINIMUM_BLOCK_RESERVED_WEIGHT.
+        if (options.block_reserved_weight && options.block_reserved_weight < MINIMUM_BLOCK_RESERVED_WEIGHT) {
```
re: https://github.com/bitcoin/bitcoin/pull/33965#discussion_r2754868641

> The interface implementation has a goal of not having implementation code in here, since we are doing this, perhaps it's time to have a wrapper for all this.

This should be happening in followup #33966 which unifies the startup & ipc checks.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/33965#pullrequestreview-3739393889

> IMO, the way the current functionality is implemented now using `std::optional` is a bit awkward, and not going to be easy to extend. for e.g if IPC has other options in the future that get set at runtime, and we don't want to override like `blockmaxweight`? Would those also need to be wrapped in `std::optional`?

It seems natural to use a std::optional type for values that `createNewBlock` callers may specify but should not be required to specify. But if you are just saying the implementation of this interface is currently awkward, that is true and followup #33966 should improve it. Additionally, we will probably want another followup PR exposing the `std::optional` capability to capnproto (https://github.com/bitcoin/bitcoin/pull/33965#pullrequestreview-3530907741), so clients can use the node `-blockreservedweight` value if desired.

> A better solution is to split `ApplyArgsManOptions` into two functions, instead `ApplyMandatoryOptions` for mandatory overrides that must be called, and `ApplyStartupOptions` for startup-specific options that only mining RPCs call, since they don't have runtime overrides. IPC would only call `ApplyMandatoryOptions`.

I think followup #33966 should do what you are describing. There is an [`ApplyArgsManOptions`](https://github.com/Sjors/bitcoin/blob/40020bc85eb252ca9a8d442999cc677dc088d94c/src/node/mining_args.cpp#L19-L45) options parsing startup options and [`ApplyMiningDefaults`](https://github.com/Sjors/bitcoin/blob/40020bc85eb252ca9a8d442999cc677dc088d94c/src/node/miner.cpp#L113-L122) applying them in the mining code.

Sjors committed:
mining: enforce minimum reserved weight for IPC

Previously a lower value was silently clamped to MINIMUM_BLOCK_RESERVED_WEIGHT.

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -125,6 +126,11 @@ def set_test_params(self):
+        # with this setting RPC templates would be empty. IPC clients set
+        # blockReservedWeight per template request and are unaffected; later in
+        # the test the IPC template includes a mempool transaction.
+        self.extra_args =[{f"-blockreservedweight={MAX_BLOCK_WEIGHT}"}, {}]
```
It's in the spirit of #34452 to continue to split this, so I added `run_ipc_option_override_test`.

Sjors commented on src/node/types.h:
```diff
@@ -44,8 +44,11 @@ struct BlockCreateOptions {
      * transaction count and coinbase transaction.
+     *
+     * Cap'n Proto IPC clients do not currently have a way of leaving this field
+     * unset and will always provide a value.
```
I updated the comment.

sedited reviewed (approved):
Re-ACK b623fab1ba87ea93dd7e302b8c927e55f2036173

ryanofsky commented on test/functional/interface_ipc_mining.py:
```diff
@@ -227,6 +228,37 @@ async def wait_for_block():
 
+    def run_ipc_option_override_test(self):
+        self.log.info("Running IPC option override test")
+        # Set an absurd reserved weight. `-blockreservedweight` is RPC-only, so
```
In commit "mining: fix -blockreservedweight shadows IPC option" (d3e49528d479613ebe50088d530a621861463fa4)

I feel like this description would be easier to understand if it began with a sentence saying what this is trying to check like "Confirm that `BlockCreateOptions.blockReservedWeight` takes precedence over `-blockreservedweight`. Set an absurdly high `-blockreservedweight` value that would result in empty blocks to verify this."

ryanofsky reviewed (approved):
Code review ACK b623fab1ba87ea93dd7e302b8c927e55f2036173. Was rebased and test split up and comment updated since last review.

ismaelsadeeq reviewed (approved):
ACK b623fab1ba87ea93dd7e302b8c927e55f2036173

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/33965#issuecomment-3601375713

> For that to work however, we'd also need to change mining.capnp, e.g.:
> 
> ```capnp
> blockReservedWeight @1 :UInt64 $Proxy.name("block_reserved_weight");
> hasBlockReservedWeight @3 :Bool = true $Proxy.skip;
> ```
> 
> Adding the `has` field and having it default to `true` means it's not a breaking change. But this doesn't actually work :-)

https://github.com/bitcoin-core/libmultiprocess/pull/243 will allow adding the "has" field so IPC clients can choose to leave this field unset and use the `-blockreservedweight` value.

I think it would be better to default the "has" field to false instead of true if added, though either value would work.

(Using std::optional and "has" fields defaulted to false could also allow simplifying the .capnp files and dropping the `.defaultBlockReservedWeight` and `.defaultCoinbaseOutputMaxAdditionalSigops` constants recently added in c6638fa7c5e97f9fd7a5ea8feb29f8caeac788bd. Probably would have been better to just make these fields optional instead of adding the default integer values at all, but both can coexist, and the default integer values could be dropped in a future version bump.)
